src/main/java/de/sldk/mc/metrics/TickDurationCollector.java

```diff
@@ -0,0 +1,54 @@
+package de.sldk.mc.metrics;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.logging.Level;
+
+import org.bukkit.Bukkit;
+import org.bukkit.Server;
+import org.bukkit.plugin.Plugin;
+
+import io.prometheus.client.Gauge;
+
+public abstract class TickDurationCollector extends Metric {
+    /*
+     * If reflection is successful, this will hold a reference directly to the
+     * MinecraftServer internal tick duration tracker
+     */
+    private static long[] tickDurationReference = null;
+
+    public TickDurationCollector(Plugin plugin, Gauge gauge, String name) {
+        super(plugin, gauge);
+
+        /*
+         * If there is not yet a handle to the internal tick duration buffer, try
+         * to acquire one using reflection.
+         *
+         * This searches for any long[] array in the MinecraftServer class. It should
+         * work across many versions of Spigot/Paper and various obfuscation mappings
+         */
+        if (tickDurationReference == null) {
+            try {
+                /* Get the actual minecraft server class */
+                Server server = Bukkit.getServer();
+                Method getServerMethod = server.getClass().getMethod("getServer");
+                Object minecraftServer = getServerMethod.invoke(server);
+
+                /* Look for the only array of longs in that class, which is tick duration */
+                for (Field field : minecraftServer.getClass().getSuperclass().getDeclaredFields()) {
+                    if (field.getType().isArray() && field.getType().getComponentType().equals(long.class)) {
+                        tickDurationReference = (long[]) field.get(minecraftServer);
```

Comment:
Extra paranoia: Do a null check on the result here? If `null` you could set the the `tickDurationReference` to an empty array - then you could get rid of the null checks in the subclasses. What do you think?

Reply:
That's a good idea, I'll do that. Another benefit of this is it won't try and fail once for each of these types - it'll just try the one time.